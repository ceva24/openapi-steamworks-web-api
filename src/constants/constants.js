const STEAM_API_URL = "https://api.steampowered.com";

export const STEAM_API_LIST_URL = `${STEAM_API_URL}/ISteamWebAPIUtil/GetSupportedAPIList/v1/`;

export const STEAM_EXTERNAL_DOCS_URL =
    "https://partner.steamgames.com/doc/webapi";

export const STEAM_PARAMETER_TYPES = {
    STRING: "string",
    UINT64: "uint64",
    UINT32: "uint32",
    INT32: "int32",
    RAWBINARY: "rawbinary",
    MESSAGE: "{message}",
};

export const SPEC_FORMATS = {
    INT32: "int32",
    INT64: "int64",
    BINARY: "binary",
    MESSAGE: "{message}",
};

export const SPEC_TYPES = {
    INTEGER: "integer",
    STRING: "string",
};

export const SPEC_VERSION = "3.0.3";

export const SPEC_INFO_TITLE = "Steamworks Web API";

export const SPEC_PATHS_PARAMETERS_IN = "query";
export const SPEC_PATHS_PARAMETERS_INPUT_JSON = "input_json";
export const SPEC_PATHS_PARAMETERS_INPUT_JSON_GET_DESCRIPTION =
    'An alternative to the query string parameters; pass arguments as a URL-encoded JSON blob. The "key" and "format" fields should still be passed as separate parameters';

export const SPEC_PATHS_PARAMETERS_INPUT_JSON_POST_DESCRIPTION =
    'An alternative to the request body; pass arguments as a URL-encoded JSON blob. The "key" and "format" fields should still be passed as separate parameters';

export const SPEC_PATHS_REQUEST_BODY_CONTENT_TYPE =
    "application/x-www-form-urlencoded";

export const SPEC_PATHS_FIELD_REQUIRED_TEXT = "(Required)";
export const SPEC_PATHS_FIELD_OPTIONAL_TEXT = "(Optional)";
export const SPEC_PATHS_FIELD_ARRAY_PARAMETER_DESCRIPTION =
    "<br>Note: this is an <a href=https://partner.steamgames.com/doc/webapi_overview#2>array parameter</a>";

export const SPEC_SERVER_URL = STEAM_API_URL;

export const SPEC_EXTERNAL_DOCS_URL = STEAM_EXTERNAL_DOCS_URL;

export const DEFAULT_RESPONSES = { 200: { description: "successful" } };

export const FORMAT_PARAMETER = {
    name: "format",
    in: SPEC_PATHS_PARAMETERS_IN,
    description: "The format of the response. Defaults to json",
    required: false,
    schema: { type: SPEC_TYPES.STRING, enum: ["json", "xml", "vdf"] },
};
