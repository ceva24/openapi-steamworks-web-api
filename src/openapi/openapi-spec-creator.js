import SwaggerParser from "@apidevtools/swagger-parser";
import {
    SPEC_SERVER_URL,
    SPEC_VERSION,
    SPEC_EXTERNAL_DOCS_URL,
    SPEC_PATHS_PARAMETERS_IN,
    SPEC_SECURITY_SCHEME_TYPE,
    SPEC_SECURITY_SCHEME_NAME,
    STEAM_API_KEY_PARAMETER,
    SPEC_PARTNER_SERVER_URL,
    SPEC_EXTERNAL_DOCS_DESCRIPTION,
} from "../constants/constants.js";
import { createInfo } from "./components/info/info-creator.js";
import { createPaths } from "./components/paths/paths-creator.js";

const createOpenApiSpec = async (apiDefinition) => {
    const openApiSpec = {
        openapi: SPEC_VERSION,
        info: createInfo(),
        servers: [{ url: SPEC_SERVER_URL }, { url: SPEC_PARTNER_SERVER_URL }],
        paths: createPaths(apiDefinition),
        externalDocs: {
            url: SPEC_EXTERNAL_DOCS_URL,
            description: SPEC_EXTERNAL_DOCS_DESCRIPTION,
        },
        security: [{ [SPEC_SECURITY_SCHEME_NAME]: [] }],
        components: {
            securitySchemes: {
                [SPEC_SECURITY_SCHEME_NAME]: {
                    type: SPEC_SECURITY_SCHEME_TYPE,
                    name: STEAM_API_KEY_PARAMETER,
                    in: SPEC_PATHS_PARAMETERS_IN,
                },
            },
        },
    };

    return SwaggerParser.validate(openApiSpec);
};

export { createOpenApiSpec };
