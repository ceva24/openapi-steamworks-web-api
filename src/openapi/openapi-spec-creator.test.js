import { SPEC_VERSION, SPEC_INFO_TITLE } from "../constants/constants.js";
import { createOpenApiSpec } from "./openapi-spec-creator.js";

const apiDefinition = {
    apilist: {
        interfaces: [
            {
                name: "IClientStats_1046930",
                methods: [
                    {
                        name: "ReportEvent",
                        version: 1,
                        httpmethod: "GET",
                    },
                ],
            },
        ],
    },
};

describe("openapi spec creator", () => {
    it("creates the openapi spec", () => {
        const openApiSpec = createOpenApiSpec(apiDefinition);

        expect(openApiSpec).toMatchSnapshot();
    });

    it("sets the version number", () => {
        const openApiSpec = createOpenApiSpec(apiDefinition);

        expect(openApiSpec.openapi).toEqual(SPEC_VERSION);
    });

    it("creates the info", () => {
        const openApiSpec = createOpenApiSpec(apiDefinition);

        expect(openApiSpec.info.title).toEqual(SPEC_INFO_TITLE);
    });

    it("creates the paths", () => {
        const openApiSpec = createOpenApiSpec(apiDefinition);

        expect(Object.keys(openApiSpec.paths).length).toEqual(1);
    });
});
