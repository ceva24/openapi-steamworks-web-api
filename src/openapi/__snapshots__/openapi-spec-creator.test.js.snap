// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openapi spec creator creates the openapi spec 1`] = `
Object {
  "externalDocs": Object {
    "url": "https://partner.steamgames.com/doc/webapi",
  },
  "info": Object {
    "title": "Steamworks Web API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/IGCVersion_1046930/GetClientVersion/v1": Object {
      "get": Object {
        "externalDocs": Object {
          "url": "https://partner.steamgames.com/doc/webapi/IGCVersion_1046930#GetClientVersion",
        },
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "description": "successful",
          },
        },
        "tags": Array [
          "IGCVersion",
        ],
      },
    },
    "/IGCVersion_1046930/GetServerVersion/v1": Object {
      "get": Object {
        "externalDocs": Object {
          "url": "https://partner.steamgames.com/doc/webapi/IGCVersion_1046930#GetServerVersion",
        },
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "description": "successful",
          },
        },
        "tags": Array [
          "IGCVersion",
        ],
      },
    },
    "/IGCVersion_1269260/GetClientVersion/v1": Object {
      "get": Object {
        "externalDocs": Object {
          "url": "https://partner.steamgames.com/doc/webapi/IGCVersion_1269260#GetClientVersion",
        },
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "description": "successful",
          },
        },
        "tags": Array [
          "IGCVersion",
        ],
      },
    },
    "/IGCVersion_1269260/GetServerVersion/v1": Object {
      "get": Object {
        "externalDocs": Object {
          "url": "https://partner.steamgames.com/doc/webapi/IGCVersion_1269260#GetServerVersion",
        },
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "description": "successful",
          },
        },
        "tags": Array [
          "IGCVersion",
        ],
      },
    },
    "/IPortal2Leaderboards_620/GetBucketizedData/v1": Object {
      "get": Object {
        "externalDocs": Object {
          "url": "https://partner.steamgames.com/doc/webapi/IPortal2Leaderboards_620#GetBucketizedData",
        },
        "parameters": Array [
          Object {
            "description": "The leaderboard name to fetch data for.",
            "in": "query",
            "name": "leaderboardName",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "successful",
          },
        },
        "tags": Array [
          "IPortal2Leaderboards",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://api.steampowered.com",
    },
  ],
}
`;
